File: map-gen-script/src/map_generator/topologies/t_shape.py
import random
from .base_topology import BaseTopology
import copy
from src.map_generator.models.path_info import PathInfo, Coord
from src.utils.geometry import add_vectors, FORWARD_X, FORWARD_Z, BACKWARD_X

class TShapeTopology(BaseTopology):
    """
    Tạo ra một con đường hình chữ T trên mặt phẳng 2D.
    Lý tưởng cho các bài học về tuần tự lệnh, hàm, hoặc các cấu trúc điều kiện
    khi người chơi phải quyết định rẽ trái hay phải ở ngã ba.
    """
    def generate_path_info_variants(self, grid_size: tuple, params: dict, max_variants: int) -> "Iterator[PathInfo]":
        """
        [KIẾN TRÚC MỚI] Tạo ra các biến thể hình chữ T với kích thước khác nhau.
        """
        count = 0
        # Lặp qua các kích thước có thể có
        for stem in range(2, 7):
            for bar in range(3, 8, 2): # Thanh ngang luôn lẻ
                if count >= max_variants: return
                
                variant_params = copy.deepcopy(params)
                variant_params['stem_length'] = stem
                variant_params['bar_length'] = bar
                yield self.generate_path_info(grid_size, variant_params)
                count += 1

    def generate_path_info(self, grid_size: tuple, params: dict) -> PathInfo:
        """
        Tạo ra một đường đi hình chữ T.

        Args:
            params (dict):
                - stem_length (int): Độ dài của "thân" chữ T.
                - bar_length (int): Độ dài của "thanh ngang" chữ T.

        Returns:
            PathInfo: Một đối tượng chứa thông tin về đường đi.
        """
        print("    LOG: Generating 't_shape' topology...")

        # Lấy độ dài các cạnh từ params, hoặc dùng giá trị ngẫu nhiên
        stem_len = params.get('stem_length', random.randint(3, 5))
        bar_len = params.get('bar_length', random.randint(4, 6))
        if bar_len % 2 == 0: bar_len += 1 # Đảm bảo thanh ngang có điểm chính giữa

        bar_side_len = bar_len // 2

        # Đảm bảo hình dạng nằm gọn trong map
        start_x = random.randint(bar_side_len + 1, grid_size[0] - bar_side_len - 2)
        start_z = random.randint(1, grid_size[2] - stem_len - 2)
        y = 0
        
        start_pos: Coord = (start_x, y, start_z)
 
        # --- PHẦN 1: TẠO HÌNH DẠNG (placement_coords) ---
        placement_coords = set()
        current_pos = start_pos
        placement_coords.add(current_pos)
 
        # 1. Vẽ thân chữ T (đi theo trục Z)
        for _ in range(stem_len):
            current_pos = add_vectors(current_pos, FORWARD_Z)
            placement_coords.add(current_pos)
 
        junction_pos = current_pos
 
        # 2. Vẽ thanh ngang (đi theo trục X)
        # Vẽ nhánh phải
        for _ in range(bar_side_len):
            current_pos = add_vectors(current_pos, FORWARD_X)
            placement_coords.add(current_pos)
 
        # Vẽ nhánh trái
        current_pos = junction_pos # Quay lại ngã ba
        for _ in range(bar_side_len):
            current_pos = add_vectors(current_pos, BACKWARD_X)
            placement_coords.add(current_pos)
 
        # --- PHẦN 2: TẠO ĐƯỜNG ĐI (path_coords) ---
        # Đường đi sẽ đi từ start_pos -> junction_pos -> cuối nhánh phải -> quay lại junction_pos -> cuối nhánh trái
        path_coords: list[Coord] = [start_pos]
        current_pos = start_pos
        
        # Đi lên thân
        for _ in range(stem_len):
            current_pos = add_vectors(current_pos, FORWARD_Z)
            path_coords.append(current_pos)
        
        # Đi sang nhánh trái (đích)
        for _ in range(bar_side_len):
            current_pos = add_vectors(current_pos, BACKWARD_X)
            path_coords.append(current_pos)

        target_pos = current_pos

        # [NEW] Tạo metadata cho Placers
        stem_coords = [start_pos]
        pos = start_pos
        for _ in range(stem_len):
            pos = add_vectors(pos, FORWARD_Z)
            stem_coords.append(pos)
        
        left_branch = [junction_pos]
        pos = junction_pos
        for _ in range(bar_side_len):
            pos = add_vectors(pos, BACKWARD_X)
            left_branch.append(pos)
        
        right_branch = [junction_pos]
        pos = junction_pos
        for _ in range(bar_side_len):
            pos = add_vectors(pos, FORWARD_X)
            right_branch.append(pos)
        
        metadata = {
            "branches": [stem_coords, left_branch, right_branch],
            "stem": stem_coords,
            "left_branch": left_branch,
            "right_branch": right_branch,
            "junction": junction_pos,
        }
 
        return PathInfo(
            start_pos=start_pos,
            target_pos=target_pos,
            path_coords=list(dict.fromkeys(path_coords)),
            placement_coords=list(dict.fromkeys(placement_coords)),
            metadata=metadata
        )

File: packages/academic-map-generator/src/generator/topologies/TShape.ts
/**
 * TShape Topology (PORTED FROM PYTHON)
 * Creates a T-shaped path with branching options
 * Ideal for conditional branching and function lessons
 */

import { BaseTopology } from './BaseTopology';
import { IPathInfo, Coord } from '../types';
import { addVectors, FORWARD_X, FORWARD_Z, BACKWARD_X } from '../utils/geometry';

export class TShapeTopology extends BaseTopology {
  *generatePathInfoVariants(
    gridSize: [number, number, number],
    params: Record<string, any>,
    maxVariants: number
  ): Generator<IPathInfo> {
    let count = 0;
    
    // Loop ranges from Python
    for (let stem = 2; stem < 7; stem++) { // 2..6
        for (let bar = 3; bar < 8; bar += 2) { // 3..7, step 2
            if (count >= maxVariants) return;
            
            const variantParams = { ...params };
            variantParams.stem_length = stem;
            variantParams.bar_length = bar;
            yield this.generatePathInfo(gridSize, variantParams);
            count++;
        }
    }
  }

  generatePathInfo(
    gridSize: [number, number, number],
    params: Record<string, any>
  ): IPathInfo {
    console.log("    LOG: Generating 't_shape' topology...");

    const stemLen = params.stem_length || (Math.floor(Math.random() * 3) + 5); // 5-7
    let barLen = params.bar_length || (Math.floor(Math.random() * 3) + 6); // 6-8
    if (barLen % 2 === 0) barLen += 1;

    const barSideLen = Math.floor(barLen / 2);
    const y = 0;

    // Start Position (Bottom of stem)
    // FIX: Use deterministic centered positioning instead of random
    // This ensures the shape stays within visible bounds and is predictable
    const centerX = Math.floor(gridSize[0] / 2);
    const centerZ = Math.floor((gridSize[2] - stemLen - 2) / 2);
    const startX = params.start_x || Math.max(barSideLen + 1, Math.min(gridSize[0] - barSideLen - 2, centerX));
    const startZ = params.start_z || Math.max(1, Math.min(gridSize[2] - stemLen - 2, centerZ));
    
    const startPos: Coord = [startX, y, startZ];

    // --- PART 1: SHAPE GENERATION ---
    const placementCoords = new Set<string>();
    const addPlacement = (c: Coord) => placementCoords.add(`${c[0]},${c[1]},${c[2]}`);
    addPlacement(startPos);
    
    let currentPos: Coord = [...startPos];

    // 1. Stem (Z+)
    for (let i = 0; i < stemLen; i++) {
        currentPos = addVectors(currentPos, FORWARD_Z);
        addPlacement(currentPos);
    }
    const junctionPos: Coord = [...currentPos];

    // 2. Bar Right (X+)
    for (let i = 0; i < barSideLen; i++) {
        currentPos = addVectors(currentPos, FORWARD_X);
        addPlacement(currentPos);
    }

    // 3. Bar Left (X-)
    currentPos = [...junctionPos];
    for (let i = 0; i < barSideLen; i++) {
        currentPos = addVectors(currentPos, BACKWARD_X);
        addPlacement(currentPos);
    }

    // --- PART 2: PATH LOGIC (SEMANTIC) ---
    
    // Reconstruct branches
    // Stem: [bottom_end ... center]
    const stemCoords: Coord[] = [startPos];
    let pos: Coord = [...startPos];
    for (let i = 0; i < stemLen; i++) {
        pos = addVectors(pos, FORWARD_Z);
        stemCoords.push(pos);
    }
    const center = stemCoords[stemCoords.length - 1]; // junction
    const bottomEnd = stemCoords[0];

    // Left Branch: [center ... left_end]
    const leftBranch: Coord[] = [center];
    pos = [...center];
    for (let i = 0; i < barSideLen; i++) {
        pos = addVectors(pos, BACKWARD_X);
        leftBranch.push(pos);
    }
    const leftEnd = leftBranch[leftBranch.length - 1];

    // Right Branch: [center ... right_end]
    const rightBranch: Coord[] = [center];
    pos = [...center];
    for (let i = 0; i < barSideLen; i++) {
        pos = addVectors(pos, FORWARD_X);
        rightBranch.push(pos);
    }
    const rightEnd = rightBranch[rightBranch.length - 1];

    const semanticPositions = {
        bottom_end: bottomEnd,
        center: center,
        left_end: leftEnd,
        right_end: rightEnd,
        optimal_start: 'bottom_end',
        optimal_end: 'right_end',
        valid_pairs: [
            {
                name: 'stem_to_right_easy',
                start: 'bottom_end',
                end: 'right_end',
                path_type: 'single_decision',
                strategies: ['conditional_branching', 'l_shape_logic'],
                difficulty: 'EASY',
                teaching_goal: 'Simple path with one direction choice'
            },
            {
                name: 'stem_to_left_medium',
                start: 'bottom_end',
                end: 'left_end',
                path_type: 'alternate_decision',
                strategies: ['conditional_branching', 'function_reuse'],
                difficulty: 'MEDIUM',
                teaching_goal: 'Alternate path with decoy on wrong branch'
            },
            {
                name: 'left_to_right_hard',
                start: 'left_end',
                end: 'right_end',
                path_type: 'cross_junction',
                strategies: ['conditional_branching', 'function_reuse'],
                difficulty: 'HARD',
                teaching_goal: 'Cross through center with items on both branches'
            }
        ]
    };

    // Helper to get Start/End logic
    // We implement `_get_start_end_positions` logic here simply:
    // Try to find valid pair matching criteria or default to optimal
    
    const allPts = [bottomEnd, center, leftEnd, rightEnd];
    let startP: Coord = bottomEnd; 
    let targetP: Coord = rightEnd;

    // Use pair logic if requested or default
    // For TShape default is usually bottom -> right
    // Python code `self._get_start_end_positions` picks from valid_pairs or randomizes
    // Since we want deterministic but rich properties, we can stick to 'optimal' default
    // or implement the full selector logic. User wants "identical logic".
    // Python `_get_start_end_positions` is in `BaseTopology`. It prioritizes `semantic_positions` if available.
    // It picks the FIRST valid pair if not specified? 
    // Actually BaseTopology logic usually defaults to "furthest apart" if no specific instruction.
    // BUT `TShape` Python calls it with `semantic_positions` dict.
    // Let's assume standard behavior: use optimal pair or first valid pair.
    
    // We will use the Optimal Pair logic as default
    startP = semanticPositions.bottom_end;
    targetP = semanticPositions.right_end;

    // Helper to get segment to center
    const getSegToCenter = (p: Coord): Coord[] => {
        const str = (c: Coord) => c.join(',');
        const pStr = str(p);
        if (pStr === str(center)) return [center];
        if (pStr === str(bottomEnd)) return stemCoords; // Bottom -> Center (Up)
        if (pStr === str(leftEnd)) return [...leftBranch].reverse() as Coord[]; // Left -> Center
        if (pStr === str(rightEnd)) return [...rightBranch].reverse() as Coord[]; // Right -> Center
        return [];
    };

    const getSegFromCenter = (p: Coord): Coord[] => {
        const str = (c: Coord) => c.join(',');
        const pStr = str(p);
        if (pStr === str(center)) return [center];
        if (pStr === str(bottomEnd)) return [...stemCoords].reverse() as Coord[]; // Center -> Bottom
        if (pStr === str(leftEnd)) return leftBranch; // Center -> Left
        if (pStr === str(rightEnd)) return rightBranch; // Center -> Right
        return [];
    };

    const pathToCenter = getSegToCenter(startP);
    const pathFromCenter = getSegFromCenter(targetP);

    let pathCoords: Coord[] = [];
    if (pathToCenter.length > 0 && pathFromCenter.length > 0) {
        pathCoords = [...pathToCenter.slice(0, -1), ...pathFromCenter];
    } else {
        pathCoords = [startP, center, targetP]; // Fallback
    }

    // Metadata
    const segments = [stemCoords, rightBranch, leftBranch];
    const corners = [center];
    const lengths = segments.map(s => s.length);

    const segment_analysis = {
        num_segments: segments.length,
        lengths: lengths,
        types: ['stem', 'branch_right', 'branch_left'],
        min_length: Math.min(...lengths),
        max_length: Math.max(...lengths),
        avg_length: lengths.reduce((a,b)=>a+b,0)/lengths.length
    };

    const metadata = {
        topology_type: "t_shape",
        semantic_positions: semanticPositions,
        branches: [stemCoords, leftBranch, rightBranch],
        corners: corners,
        stem: stemCoords,
        left_branch: leftBranch,
        right_branch: rightBranch,
        junction: center,
        segments: segments,
        segment_analysis: segment_analysis
    };

    const placementList: Coord[] = [];
    placementCoords.forEach(k => placementList.push(k.split(',').map(Number) as Coord));

    return {
        start_pos: startP,
        target_pos: targetP,
        path_coords: pathCoords,
        placement_coords: placementList,
        obstacles: [],
        metadata: metadata
    };
  }
}
